proposals/p5914.md

```diff
@@ -0,0 +1,277 @@
+    aliasing.
+-   More incremental enforcement of data-race safety.
+-   Aligning concepts and idioms with C++.
+-   Having fewer assumptions about unsafe code.
+
+First, Carbon is going to include language constructs to match C++ that Rust
+avoids completely, and this comes at a complexity cost. A canonical example is
+inheritance: Carbon will have inheritance and virtual dispatch built into the
+language, allowing straightforward C++ interop and migration for APIs that use
+it. These features increase the size of the language overall, and through
+interaction increase the complexity of safety features.
+
+The second of these is the most significant change to safety: supporting safe,
+shared mutation. We think this will open up the door to significantly more
+flexible code while still providing memory safety. With Rust, there are two
+kinds of safe "pointers": shared borrows (`&`), and mutable borrows (`&mut`).
+The latter are required to be exclusive -- no other safe "pointers" alias the
+mutable object. We expect to have more distinguished types of "pointers" in
+Carbon to be able to model more C++ coding patterns safely, including both
+shared mutable "pointers", and exclusive mutable pointers -- this is where the
+complexity increases. While the exact design of this needs to be carefully
+fleshed out, our current idea is to build on the concepts of
+[safe, shared mutability in the Ante Language](https://antelang.org/blog/safe_shared_mutability/).
+Building illustrative examples of how this impacts C++ code migrating towards
+memory safety is an important part of our planned deliverables for 2025.
+
+Third, we expect to have more incremental enforcement of data-race safety in
+Carbon, similar to how
+[Swift allowed code to incrementally adopt the necessary patterns and restrictions](https://www.swift.org/documentation/concurrency/)
+to be data-race safe. The necessary infrastructure was provided as independently
+adoptable APIs, and enforcement was then available optionally prior to Swift 6
+making this a requirement.
+
+Fourth, we will express Carbon's core concepts around mutation in terms of C++
+idioms and concepts. For example, in a container, we expect iterators to use
+shared mutable pointers, and the C++ concept of "iterator invalidation" will be
+captured by methods that require exclusive mutable access, precluding any
+outstanding shared mutable pointers.
+
+Lastly, Carbon's safety model will make fewer assumptions about what unsafe code
+is allowed to do than Rust. To avoid undefined behavior, the optimization of
+safe code will not rely on properties that are not established locally. For
+example, there won't be an assumption that a call to C++ code won't save a copy
+of pointers to passed-in objects. When there are restrictions on what unsafe
+code can do, we will endeavor to make violations into
+[erroneous behavior](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2795r2.html#bigpic),
```

Comment:
Note somewhere this should be updated once we've checked in a definition of erroneous behavior?

Reply:
This is in the proposal, so wouldn't expect this to be updated? This is linking to the *origins* of the suggestion to define erroneous behavior, and that's really the WG21 paper.